src/controller/proxy/controller.go

```diff
@@ -0,0 +1,224 @@
+	ProxyBlob(ctx context.Context, p *models.Project, repo string, dig string, w http.ResponseWriter, r RemoteInterface) error
+	// ProxyManifest proxy the manifest to the target server
+	ProxyManifest(ctx context.Context, p *models.Project, repo string, art lib.ArtifactInfo, w http.ResponseWriter, r RemoteInterface) error
+}
+
+type controller struct {
+	blobCtl     blob.Controller
+	registryMgr registry.Manager
+	mu          sync.Mutex
+	inflight    map[string]interface{}
+	artifactCtl artifact.Controller
+	local       LocalInterface
+}
+
+func (c *controller) isProxyReady(p *models.Project) bool {
+	if p.RegistryID < 1 {
+		return false
+	}
+	reg, err := c.registryMgr.Get(p.RegistryID)
+	if err != nil {
+		log.Errorf("failed to get registry, error:%v", err)
+		return false
+	}
+	return reg.Status == model.Healthy
+}
+
+func (c *controller) UseLocalManifest(ctx context.Context, p *models.Project, art lib.ArtifactInfo) bool {
+	if !c.isProxyReady(p) {
+		return true
+	}
+	if len(string(art.Digest)) > 0 {
+		exist, err := c.local.BlobExist(ctx, art.Digest)
+		if err == nil && exist {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *controller) UseLocalBlob(ctx context.Context, p *models.Project, digest string) bool {
+	if !c.isProxyReady(p) {
+		return true
+	}
+	exist, err := c.local.BlobExist(ctx, digest)
+	if err != nil {
+		return false
+	}
+	return exist
+}
+
+func (c *controller) ProxyManifest(ctx context.Context, p *models.Project, repo string, art lib.ArtifactInfo, w http.ResponseWriter, r RemoteInterface) error {
+	var man distribution.Manifest
+	var err error
+	ref := string(art.Digest)
+	if len(ref) == 0 {
+		ref = art.Tag
+	}
+	if len(ref) == 0 {
+		ref = "latest"
+	}
+	man, err = r.Manifest(repo, ref)
+	if err != nil {
+		if errors.IsNotFoundErr(err) {
+			go func() {
+				c.local.DeleteManifest(ctx, repo, string(art.Tag))
+			}()
+		}
+		return err
+	}
+
+	ct, payload, err := man.Payload()
+	setHeaders(w, int64(len(payload)), ct, art.Digest)
+	w.Write(payload)
+
+	// Push manifest in background
+	go func() {
+		c.waitAndPushManifest(ctx, p, repo, string(art.Tag), man, art, ct, r)
+	}()
+
+	return nil
+}
+
+func (c *controller) ProxyBlob(ctx context.Context, p *models.Project, repo string, dig string, w http.ResponseWriter, r RemoteInterface) error {
+	log.Debugf("The blob doesn't exist, proxy the request to the target server, url:%v", repo)
+	desc := distribution.Descriptor{}
+	size, bReader, err := r.BlobReader(repo, dig)
+	if err != nil {
+		log.Errorf("failed to pull blob, error %v", err)
+		return err
+	}
+	defer bReader.Close()
+	written, err := io.CopyN(w, bReader, size)
+	if written != size {
+		e := errors.Errorf("The size mismatch, actual:%d, expected: %d", written, size)
+		return e
+	}
+	desc.Size = size
+	desc.Digest = digest.Digest(dig)
+
+	setHeaders(w, size, desc.MediaType, dig)
+	go func() {
+		err := c.putBlobToLocal(ctx, p, repo, p.Name+"/"+repo, desc, r)
```

Comment:
Put this before `written, err := io.CopyN(w, bReader, size)`  so that the two process can run concurrently